src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryDataLoader.kt

```diff
@@ -407,6 +438,219 @@ object ChocolateFactoryDataLoader {
+        if (bestUpgrade.isRabbit && bestUpgrade.level == 19) {
+            val nextRabbit = getNextRabbit(bestUpgrade)
+            if (nextRabbit != null) {
+                list.add(nextRabbit)
+                allUpgrades[nextRabbit.slotIndex] = LinkedList()
+            }
+        }
+
+        // Update extra per second and effective costs for all current upgrades.
+        updateEffectiveCost(list, nextBaseIncrease, nextMultiplierIncrease)
+
+        // recursive call to find the next best upgrade after spending the cost of the best upgrade.
+        return findAllBestUpgradesImpl(
+            list,
+            allUpgrades,
+            remainingChocolate - (bestUpgrade.price ?: Long.MAX_VALUE),
+            nextBaseIncrease,
+            nextMultiplierIncrease,
+        )
+    }
+
+    /** Calculates the upgrade Cost for the upgrade at [slotIndex] with [level].
+     *
+     * @param slotIndex the slot index of the upgrade
+     * @param level the level of the upgrade
+     *
+     * @return the cost of the upgrade || null if the upgrade is maxed.
+     */
+    private fun getUpgradeCost(slotIndex: Int, level: Int): Long? {
+        var price: Long? = null
+        if (level < (ChocolateFactoryAPI.maxUpgradeLevelPerPrestige[slotIndex]?.getOrNull(
+                ChocolateFactoryAPI.currentPrestige - 1,
+            ) ?: 0)
+        ) {
+
+            // Use upgrade cost per level if it exists, otherwise use the formula.
+            if ((ChocolateFactoryAPI.upgradeCostPerLevel[slotIndex]?.size ?: 0) > level) {
+                val nextRaw = ChocolateFactoryAPI.upgradeCostPerLevel[slotIndex]?.get(level) ?: 0
+                val prestigeMultiplier =
+                    (ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("prestige")
+                        ?: 0.0) * (ChocolateFactoryAPI.currentPrestige - 1)
+                price = floor(nextRaw * (1 + prestigeMultiplier)).toLong()
+            } else {
+                val base = ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("base") ?: 0.0
+                val multiplier = ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("exp") ?: 0.0
+                val prestigeMultiplier =
+                    (ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("prestige")
+                        ?: 0.0) * (ChocolateFactoryAPI.currentPrestige - 1)
+                price = floor(round(base * multiplier.pow((level + 1)) * (1 + prestigeMultiplier))).toLong()
+            }
+        }
+        return price
+    }
+
+    /** Constructs the rabbit with level = 0 after [upgrade] if upgrade is a rabbit.
+     *  Does not recalculate [ChocolateFactoryUpgrade.effectiveCost] and [ChocolateFactoryUpgrade.extraPerSecond] but
+     *  instead just copies it from [upgrade]
+     *
+     * @param upgrade the current upgrade
+     */
+    private fun getNextRabbit(upgrade: ChocolateFactoryUpgrade): ChocolateFactoryUpgrade? {
```

Comment:
functions like this do not belong into a "data loader" class

Reply:
Should it be moved to the API class then? Or just the factory upgrade class? I originally put it there because only `ChocolateFactoryDataLoader.findAllBestUpgradesImpl` depends on it but I see where that isn't enough justification to put it there.

Reply:
`ChocolateFactoryAPI` already gets quite full. Ideally, we would entirely restructure each function over all classes in the chocolate package. But for the scope of this pr, I, think it's fine when you move the new functions either into existing utils classes or create a new RabbitUtils class. Up to you.